plutus-core/executables/pir/Main.hs

```diff
@@ -1,117 +1,242 @@
 import PlutusIR.Compiler qualified as PIR
+import PlutusIR.Core.Instance.Pretty ()
 import PlutusIR.Core.Plated
 import PlutusPrelude
+import UntypedPlutusCore qualified as UPLC
+
+import Control.Lens (coerced, (^..))
+import Control.Monad (when)
+import Control.Monad.Trans.Except (runExcept)
+import Control.Monad.Trans.Reader (runReader, runReaderT)
+import Data.ByteString.Lazy.Char8 qualified as BSL
+import Data.Csv qualified as Csv
+import Data.IntMap qualified as IM
+import Data.List (sortOn)
+import Data.Text qualified as T
+import Options.Applicative
 import Text.Megaparsec (errorBundlePretty)
 
-type PLCTerm  = PLC.Term PLC.TyName PLC.Name PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
-type PIRError = PIR.Error PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
-type PIRCompilationCtx a = PIR.CompilationCtx PLC.DefaultUni PLC.DefaultFun a
-data Command = Analyse IOSpec
-             | Compile COpts
-             | Print PrintOptions
-
-data COpts = COpts
-  { cIn       :: Input
-  , cOptimize :: Bool
-  }
-
-pCOpts :: Parser COpts
-pCOpts = COpts
-            <$> input
-            <*> switch' ( long "dont-optimize"
-                        <> help "Don't optimize"
-                        )
-  where
-    switch' :: Mod FlagFields Bool -> Parser Bool
-    switch' = fmap not . switch
-
-pPirOpts :: Parser Command
-pPirOpts = hsubparser $
-    command "analyse"
-        (info (Analyse <$> ioSpec) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise and analyse the program, " <>
-              "looking for variables with the largest retained size.")
-  <> command "compile"
-        (info (Compile <$> pCOpts) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise it, " <>
-              "and test if it can be successfully compiled to PLC.")
-  <> command "print"
-        (info (Print <$> printOpts) $
-            progDesc $
-              "Given a PIR program in flat format, " <>
-              "deserialise it and print it out textually.")
-
--- | Load flat pir and deserialize it
-loadPir :: Input -> IO (PirProg ())
-loadPir = loadASTfromFlat Named
-
-compile :: COpts -> PirProg () -> Either PIRError PLCTerm
-compile opts (PIR.Program _ pirT) = do
+type PirError a = PIR.Error PLC.DefaultUni PLC.DefaultFun a
+type UnitProvenance = PIR.Provenance ()
+
+
+{- Note [De Bruijn indices and PIR]
+   The `plc` and `uplc` commands both support ASTs whose "names" are de Bruijn
+   inidces.  These aren't supported for PIR because PIR has `Let` blocks of
+   possibly mutually recursive definitions and it's not clear whether de Bruijn
+   indices (or levels) really make sense in the presence of mutual recursion
+   since scopes aren't properly nested in that case: if we process the AST in
+   the normal way then a declaration may have to refer to another variable which
+   hasn't yet been declared.  It may be possbile to overcome this by allowing
```

Comment:
typo `possible`

Reply:
Thanks.  GitHub suggestions (the +/ thing above the comment window) are good for typos!